docs/src/diagnostics.md

```diff
@@ -0,0 +1,277 @@
+provided by `ClimaAtmos` use this metadata.
+
+In `ClimaAtmos`, we follow the convention that:
+
+- `short_name` is the name used to identify the variable in the output files and
+                in the file names. It is short, but descriptive. We identify
+                diagnostics by their short name, so the diagnostics defined by
+                `ClimaAtmos` have to have unique `short_name`s. We follow the
+                Coupled Model Intercomparison Project (CMIP) conventions.
+
+- `long_name`: Name used to describe the variable in the output file.
+
+- `units`: Physical units of the variable.
+
+- `comments`: More verbose explanation of what the variable is, or comments related to how
+              it is defined or computed.
+
+In `ClimaAtmos`, we try to follow [this Google
+spreadsheet](https://docs.google.com/spreadsheets/d/1qUauozwXkq7r1g-L4ALMIkCNINIhhCPx)
+for variable naming, with a `long_name` the does not have spaces and capital
+letters. [Standard
+names](http://cfconventions.org/Data/cf-standard-names/current/build/cf-standard-name-table.html)
+are not used.
+
+### Compute function
+
+The other piece of information needed to specify a `DiagnosticVariable` is a
+function `compute_from_integrator!`. Schematically, a `compute_from_integrator!` has to look like
+```julia
+function compute_from_integrator!(out, integrator)
+    # FIXME: Remove this line when ClimaCore implements the broadcasting to enable this
+    out .= # Calculcations with the state (= integrator.u) and the parameters (= integrator.p)
+end
+```
+Diagnostics are implemented as callbacks function which pass the `integrator`
+object (from `OrdinaryDiffEq`) to `compute_from_integrator!`.
+
+`compute_from_integrator!` also takes a second argument, `out`, which is used to
+avoid extra memory allocations (which hurt performance). If `out` is `nothing`,
+and new area of memory is allocated. If `out` is a `ClimaCore.Field`, the
+operation is done in-place without additional memory allocations.
+
+If your diagnostic depends on the details of the model, we recommend using
+additional functions so that the correct one can be found through dispatching.
+For instance, if you want to compute relative humidity, which does not make
+sense for dry simulations, you should define the functions
+
+```julia
+function compute_relative_humidity_from_integrator!(
+    out,
+    integrator,
+    moisture_model::T,
+) where {T}
+    error("Cannot compute relative_humidity with moisture_model = $T")
+end
+
+function compute_relative_humidity_from_integrator!(
+    out,
+    integrator,
+    moisture_model::T,
+) where {T <: Union{EquilMoistModel, NonEquilMoistModel}}
+    # FIXME: Avoid extra allocations when ClimaCore overloads .= for this use case
+    # We will want: out .= integrator.u.c.ρ
+    thermo_params = CAP.thermodynamics_params(integrator.p.params)
+    return TD.relative_humidity.(thermo_params, integrator.p.ᶜts)
```

Comment:
Can we address these TODOs before merging? Fixing this may have a big impact on the overall design.

Reply:
```suggestion
    # FIXME: Avoid extra allocations when ClimaCore overloads .= for this use case
    # We will want: out .= integrator.u.c.ρ
    thermo_params = CAP.thermodynamics_params(integrator.p.params)
    if isnothing(out)
        return TD.relative_humidity.(thermo_params, integrator.p.ᶜts)
    else
        return out .= TD.relative_humidity.(thermo_params, integrator.p.ᶜts)
    end
```